src/drivers/fluid_wasapi.c

```diff
@@ -181,60 +176,182 @@ fluid_audio_driver_t *new_fluid_wasapi_audio_driver2(fluid_settings_t *settings,
+        goto cleanup;
+    }
 
     if(fluid_settings_str_equal(settings, "audio.sample-format", "16bits"))
     {
-        sample_size = sizeof(int16_t);
-        wfx.Format.wFormatTag = WAVE_FORMAT_PCM;
+        dev->sample_size = sizeof(int16_t);
+        dev->sample_format = WAVE_FORMAT_PCM;
     }
     else
     {
-        sample_size = sizeof(float);
-        wfx.Format.wFormatTag = WAVE_FORMAT_IEEE_FLOAT;
+        dev->sample_size = sizeof(float);
+        dev->sample_format = WAVE_FORMAT_IEEE_FLOAT;
     }
 
-    wfx.Format.nChannels  = 2;
-    wfx.Format.wBitsPerSample  = sample_size * 8;
-    wfx.Format.nBlockAlign     = sample_size * wfx.Format.nChannels;
-    wfx.Format.nSamplesPerSec  = (DWORD)sample_rate;
-    wfx.Format.nAvgBytesPerSec = (DWORD)sample_rate * wfx.Format.nBlockAlign;
-    //wfx.Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
-    //wfx.Format.cbSize = 22;
-    //wfx.SubFormat = guid_float;
-    //wfx.dwChannelMask = SPEAKER_FRONT_LEFT | SPEAKER_FRONT_RIGHT;
-    //wfx.Samples.wValidBitsPerSample = wfx.Format.wBitsPerSample;
-
-    dev = FLUID_NEW(fluid_wasapi_audio_driver_t);
-
-    if(dev == NULL)
+    if(fluid_settings_dupstr(settings, "audio.wasapi.device", &dev->dname) != FLUID_OK)
     {
         FLUID_LOG(FLUID_ERR, "wasapi: out of memory.");
-        return NULL;
+        goto cleanup;
     }
 
-    FLUID_MEMSET(dev, 0, sizeof(fluid_wasapi_audio_driver_t));
-
     dev->func = func;
     dev->user_pointer = data;
-    dev->buffer_duration = periods * period_size / sample_rate;
-    dev->channels_count = audio_channels * 2;
-    dev->float_samples = (wfx.Format.wFormatTag == WAVE_FORMAT_IEEE_FLOAT);
-    buffer_duration_reftime = (fluid_long_long_t)(dev->buffer_duration * 1e7 + .5);
-    periods_reftime = (fluid_long_long_t)(period_size / sample_rate * 1e7 + .5);
+    dev->buffer_duration = dev->periods * dev->period_size / dev->sample_rate;
+    dev->channels_count = dev->audio_channels * 2;
+    dev->float_samples = (dev->sample_format == WAVE_FORMAT_IEEE_FLOAT);
+    dev->buffer_duration_reftime = (fluid_long_long_t)(dev->buffer_duration * 1e7 + .5);
+    dev->periods_reftime = (fluid_long_long_t)( dev->period_size / dev->sample_rate * 1e7 + .5);
+
+    dev->quit_ev = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    if(dev->quit_ev == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: failed to create quit event.");
+        goto cleanup;
+    }
+
+    dev->start_ev = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    if(dev->start_ev == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: failed to create start event.");
+        goto cleanup;
+    }
+
+    dev->thread = CreateThread(NULL, 0, fluid_wasapi_audio_run, (void *)dev, 0, &dev->thread_id);
+
+    if(dev->thread == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: failed to create audio thread.");
+        goto cleanup;
+    }
+
+    if (WaitForMultipleObjects(2, &dev->start_ev, FALSE, 200) == WAIT_OBJECT_0) {
```

Comment:
Please do not rely on the order of struct members. If we add a member after `start_ev` this will horribly break. Pls. use an explicit array. Also, I'd vote to increase the timeout to 1000 msec. Ideally, you could add a
`FLUID_LOG(FLUID_WARN, "wasapi: initialization timeout!")` if `WAIT_TIMEOUT` was returned.